### PR TITLE
Testing functionality

### PR DESCRIPTION
**Pull request by udit-gulati**

### CODE REVIEW

File: packages/polar/sample-project/test/sample-test.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
+const { expect, use } = require("chai");
+const { Contract, getAccountByName, polarChai } = require("secret-polar");
+
+use(polarChai);
```
Comment (coco-sha) [inline packages/polar/sample-project/test/sample-test.js:4]: why do we use this line ? does user needs to include it everytime ?

File: packages/polar/sample-project/test/sample-test.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+  const deploy_response = await contract.deploy(contract_owner);
+  console.log(deploy_response);
+
+  const contract_info = await contract.instantiate({"count": 102}, "deploy test", contract_owner);
+  console.log(contract_info);
+
+  await expect(contract.tx.reset(other, [], 100)).to.be.revertedWith("unauthorized");
+
+  await expect(contract.query.get_count()).to.respondWith({ 'count': 100 });
```
Comment (coco-sha) [inline packages/polar/sample-project/test/sample-test.js:20]: could you please add a test where we increment count and check if it is incremented ?

File: packages/polar/src/builtin-tasks/test.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+      runtimeEnv
+    );
+  }
+}
+
+async function executeTestTask (
+  { tests }: Input,
+  runtimeEnv: PolarRuntimeEnvironment
+  // eslint-disable-next-line
+): Promise<any> {
```
Comment (coco-sha) [inline packages/polar/src/builtin-tasks/test.ts:41]:
```suggestion
): Promise<void> {
```

Comment (coco-sha) [reply packages/polar/src/builtin-tasks/test.ts:41]: we can remove the above line

File: packages/polar/src/lib/chai/matchers/changeScrtBalance.ts [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    // raise exception, should be function
+  }
+
+  const client = getClient(PolarContext.getPolarContext().getRuntimeEnv().network);
+
+  const balanceBefore = extractScrtBalance(
+    (await client.getAccount(accountAddr) as WasmAccount).balance
+  );
+
+  const txResponse = await transaction();
```
Comment (coco-sha) [inline packages/polar/src/lib/chai/matchers/changeScrtBalance.ts:66]: unused variable

File: packages/polar/src/lib/chai/matchers/changeScrtBalance.ts [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+  const balanceAfter = extractScrtBalance(
+    (await client.getAccount(accountAddr) as WasmAccount).balance
+  );
+
+  // if (
```
Comment (coco-sha) [inline packages/polar/src/lib/chai/matchers/changeScrtBalance.ts:72]: do we need lines below ?

File: packages/polar/src/lib/chai/matchers/changeTokenBalance.ts [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+  return 0;
+}
+
+export async function getBalanceChange (
+  transaction: (() => Promise<any>), // eslint-disable-line  @typescript-eslint/no-explicit-any
+  accountAddr: string,
+  token: string
+): Promise<number> {
+  if (typeof transaction !== 'function') {
+    // raise exception, should be function
```
Comment (coco-sha) [inline packages/polar/src/lib/chai/matchers/changeTokenBalance.ts:58]: raise

File: packages/polar/src/lib/chai/matchers/changeTokenBalance.ts [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+  }
+
+  const client = getClient(PolarContext.getPolarContext().getRuntimeEnv().network);
+
+  const balanceBefore = extractTokenBalance(
+    (await client.getAccount(accountAddr) as WasmAccount).balance,
+    token
+  );
+
+  const txResponse = await transaction();
```
Comment (coco-sha) [inline packages/polar/src/lib/chai/matchers/changeTokenBalance.ts:68]: unused

File: packages/polar/src/lib/chai/matchers/changeTokenBalances.ts [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+  );
+}
+
+export async function getBalanceChanges (
+  transaction: (() => Promise<any>), // eslint-disable-line  @typescript-eslint/no-explicit-any
+  accountAddresses: string[],
+  token: string
+): Promise<number[]> {
+  if (typeof transaction !== 'function') {
+    // raise exception, should be function
```
Comment (coco-sha) [inline packages/polar/src/lib/chai/matchers/changeTokenBalances.ts:78]: ^

File: packages/polar/src/lib/chai/matchers/changeTokenBalances.ts [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+  }
+
+  const balancesBefore = await getBalances(accountAddresses, token);
+
+  const txResponse = await transaction();
```
Comment (coco-sha) [inline packages/polar/src/lib/chai/matchers/changeTokenBalances.ts:83]: unused

File: packages/polar/src/lib/deploy/contract.ts [hunk 10]
```diff
@@ -187,6 +212,17 @@ export class Contract {
...
+    // for (const responseAbi of this.responseAbis) {
+    //   for (const message of responseAbi.messages) {
+    //     const msgName: string = message.identifier;
+    //     const args: AbiParam[] = message.args;
+
+    //     if (this.responses[msgName] == null) {
+    //       this.responses[msgName] = args;
+    //     }
+    //   }
+    // }
```
Comment (coco-sha) [inline packages/polar/src/lib/deploy/contract.ts:225]: what is this for ?

File: packages/polar/sample-project/test/sample-test.js (hunk 1 above)
Comment (udit-gulati) [reply packages/polar/sample-project/test/sample-test.js:4]: yes, this is required here. Other option was to add this in the library, but there the path of chai module was to be defined as to separate it from chai used for testing the library. One more way is to use this line in polar config as like used in hardhat but this is cleaner.

File: packages/polar/sample-project/test/sample-test.js [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+  console.log(contract_info);
+
+  await expect(contract.tx.reset(other, [], 100)).to.be.revertedWith("unauthorized");
+
+  await expect(contract.query.get_count()).to.respondWith({ 'count': 102 });
+
+  await expect(contract.query.get_count()).to.respondWith({ 'count': 1000 }); // this will fail
+}
+
+module.exports = { default: run };
```
Comment (coco-sha) [inline packages/polar/sample-project/test/sample-test.js:25]: could you please update this file similar to mocha testing ?

`describe("counter example")`
`it("should incrment counter").....`

File: packages/polar/src/lib/chai/matchers/changeScrtBalance.ts [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+
+import { PolarContext } from "../../../internal/context";
+import { PolarError } from "../../../internal/core/errors";
+import { ERRORS } from "../../../internal/core/errors-list";
+import type {
+  Account, Coin, UserAccount
+} from "../../../types";
+import { getClient } from "../../client";
+
+const defaultFees = {
```
Comment (coco-sha) [inline packages/polar/src/lib/chai/matchers/changeScrtBalance.ts:12]: could you please move this to constants file ?

File: packages/polar/sample-project/test/sample-test.js (hunk 2 above)
Comment (udit-gulati) [reply packages/polar/sample-project/test/sample-test.js]: done

**Review by coco-sha — APPROVED** [review]
looks good, Please remove linter errors

**Pull request merged**